### ISSUE TITLE
Incorrect namespace for generated types

### ISSUE THREAD

**Issue by rkevingibson** [microsoft/win32metadata#1639]
I'm working on rust bindings for the DirectX Agility SDK, which you can find here: https://github.com/rkevingibson/agilitysdk_rs. I'm generally following [this blog from Rafael Rivera](https://withinrafael.com/2023/01/18/generating-metadata-for-the-windows-crate/) to guide me through it. 
One issue I'm having is with two function arguments where the types that are generated are pointing to the types in the WinSDK rather than the ones with the same names in the headers I am parsing - in particular, two function pointers.

For example, in the headers I'm scraping there is a type defined, `D3D12MessageFunc` which is a function pointer. It gets passed as an argument to `ID3D12InfoQueue1::RegisterMessageCallback()`. However, in the generated rust bindings, the type in the function argument is instead `::windows::Win32::Graphics::Direct3D12::D3D12MessageCallback` - instead of the version of the type that is defined locally, it points to the version in the WinSDK. I suspect the actual issue is in the WinMD generation, as when I look at `RegisterMessageCallback()` in ILSpy, I see a `using Windows.Win32.Graphics.Direct3D12;`, suggesting the type is already mapped to the WinSDK version. 

The same issue happens with another function pointer, `PFN_DESTRUCTION_CALLBACK`.

I've tried removing the default scraper.header.txt and making sure there's no reference to D3D12 but that didn't seem to do anything. I've also tried playing with functionPointerFixups.json, but no luck. In both cases the types in WinSDK and AgilitySDK match, so it's not a big deal, but for the generated bindings it means pulling in effectively a redundant version of the D3D API which I would prefer to avoid if possible.

**Comment by riverar** [comment]
Hey @rkevingibson 👋. The Direct3D types in win32metadata are actually sourced from the Agility SDK (1.610.0 as of writing, https://github.com/microsoft/win32metadata/pull/1445). And I believe that metadata has been incorporated into the latest `windows` crate. Does that simplify your efforts perhaps?

(We'll get that updated to 1.610.4.)

If you're trying to bind against the _preview_ Agility SDK, you'll need to generate metadata (as you've already done) and depending on your crate, switch off the Direct3D `windows` crate features or generate your bindings in a completely different namespace.

(Thanks for reminding me that I need to update that blog post to take advantage of the new windows-rs _riddle_ tool for binding generation.)

**Comment by rkevingibson** [comment]
Yes, I'm trying to bind against the preview SDK. I've generated the bindings in a unique namespace, and am trying to compile them without the Direct3D windows crate features at all - this last step is where the issue lies. When compiling the generated without those features, the compilation will fail, because certain types are specified as being from the `windows` crate's namespace, and not the namespace I'm defining for the bindings, even though the same type exists.

E.g. I see this function signature:
```rust
pub unsafe fn RegisterMessageCallback(
        &self,
        callbackfunc: ::windows::Win32::Graphics::Direct3D12::D3D12MessageFunc,
        callbackfilterflags: D3D12_MESSAGE_CALLBACK_FLAGS,
        pcontext: *mut ::core::ffi::c_void,
        pcallbackcookie: *mut u32,
    )
```
which requires the `Win32_Graphics_Direct3D12` feature to compile. I would prefer it emit
```rust
pub unsafe fn RegisterMessageCallback(
        &self,
        callbackfunc: D3D12MessageFunc,
        callbackfilterflags: D3D12_MESSAGE_CALLBACK_FLAGS,
        pcontext: *mut ::core::ffi::c_void,
        pcallbackcookie: *mut u32,
    )
```
which is the same, except for the type of the `callbackfunc` parameter. The `D3D12MessageFunc` type is defined elsewhere in the same file, so the type is generated but not used for some reason.

**Comment by mikebattista** [comment]
See https://github.com/microsoft/win32metadata/pull/1597 for some discussion on generating bindings for the preview SDK.

@MarijnS95 can probably help you.

**Comment by riverar** [comment]
Re-opening this because I ran through Kevin's repro and I see the `CallbackFunc` is tied to a type in `Windows.Win32.winmd`.

```csharp
.method public hidebysig newslot abstract virtual 
	instance valuetype [Windows.Win32.winmd]Windows.Win32.Foundation.HRESULT RegisterMessageCallback (
		[in] class [Windows.Win32.winmd]Windows.Win32.Graphics.Direct3D12.D3D12MessageFunc CallbackFunc,
		[in] valuetype [Microsoft.AgilitySdk.winmd]Microsoft.AgilitySdk.D3D12_MESSAGE_CALLBACK_FLAGS CallbackFilterFlags,
		[in] [out] void* pContext,
		[in] [out] uint32* pCallbackCookie
	) cil managed 
```

This is unexpected because we're generating `Microsoft.AgilitySdk.ID3D12InfoQueue1` and `D3D12MessageFunc` is a local type. I think the generator should give preference to local types for the scenario where folks want to augment/override metadata.

**Comment by riverar** [comment]
To explore the idea, I implemented a `--ignoreTypeExpressions` parameter as a POC: https://github.com/microsoft/win32metadata/compare/main...riverar:win32metadata:rafael/ignoreTypeExpressions?expand=1

This adds the ability to filter out types using regular expressions, as such:
```
--ignoreTypeExpressions
Windows\.Win32\.Graphics\.Direct3D12\.
```

Which then results in an interface generated with local types:
```diff
.method public hidebysig newslot abstract virtual 
  instance valuetype [Windows.Win32.winmd]Windows.Win32.Foundation.HRESULT RegisterMessageCallback (
+   [in] class Microsoft.AgilitySdk.D3D12MessageFunc CallbackFunc,
    [in] valuetype [Microsoft.AgilitySdk.winmd]Microsoft.AgilitySdk.D3D12_MESSAGE_CALLBACK_FLAGS CallbackFilterFlags,
    [in] [out] void* pContext,
    [in] [out] uint32* pCallbackCookie
  ) cil managed 
```

If we continue forward with this, we should rename the option and make it easier to use. That is, I think it should accept a type name string by default and only build up a regular expression if requested. (Perhaps with a `regex:` prefix?)

**Comment by MarijnS95** [comment]
> I'm working on rust bindings for the DirectX Agility SDK

Oh... You don't need to do any of that by hand as it has all been included in win32metadata and windows-rs quite some time ago. For the preview bindings one simply needs to update and regenerate the two:

https://github.com/microsoft/win32metadata/pull/1597
https://github.com/Traverse-Research/windows-rs/compare/regen-agility-711-preview

If you however cannot replace your `windows-rs` crates with this git version / fork, I have previously gone the same route of creating separate metadata bindings and a Rust crate for the `DirectX-Headers` repo (that sparked the idea to consume the bindings directly in win32metadata, with the massive downside that the preview release mechanism is now completely broken):

https://github.com/microsoft/DirectX-Headers/pull/82
https://github.com/MarijnS95/DirectX-Headers/compare/main
https://crates.io/crates/microsoft-directx
https://crates.io/crates/microsoft-direct3d

Those will need updating to the latest AgilitySDK drop if you wish to use them, but it's only a temporary workaround. The setup looks very similar to the repo you linked, so we've all been doing duplicate work :sweat_smile: 

It should also be much easier with all the `windows-core` work going around lately, and the `riddle` tool. @rkevingibson perhaps using the latest metadata tooling implicitly solves some of these problems? `0.48` is already 4 months old, and `windows-bindgen` is no more.

---

This ties in perfectly to what I said in https://github.com/microsoft/win32metadata/pull/1597 about visibility and _considering the metadata and Rust a first-class citizen when it comes to preview releases_ (of the Agility SDK and perhaps other software stacks). There's clearly a demand for consuming the latest preview bindings on "day one" in something that's not C++.
